### PR TITLE
Update to .NET 8 and leverage keyed services for decoration

### LINKED ISSUE #65: AsImplementedInterfaces() registering too much
I understand that AsImplementedInterfaces() registers *all* implemented interfaces, however this is sometimes too much. Take for instance, the FluentValidation library, you may scan with the following:

`.AddClasses(c => c.AssignableTo(typeof(IValidator<>))).AsImplementedInterfaces()`

This will pickup all the `AbstractValidator<T>` implementations you have, but this class implements the following interfaces:
```
IValidator<T>
IEnumerable
IEnumerable<IValidationRule>
IValidator
```

Which means you get entries for all of these, not just the desired `IValidator<T>` and I need to do the following to get rid of the superfluous entries:

```
services.RemoveAll<IValidator>();
services.RemoveAll<IEnumerable>();
services.RemoveAll<IEnumerable<IValidationRule>>();
```

Is there any way to do this without having to remove the extra entries afterwards?

### LINKED ISSUE #171: Azure Function Decoration issues in 4.2.0
Using Scrutor 4.2.0 with Azure Functions seems to cause issues.

Error when running the function (locally and remotely):
"Microsoft.Azure.WebJobs.Script.WebHost: Registering implementation type FunctionApp1.BasicService is not assignable to service type FunctionApp1.IService."

When downgrading to version 4.1.0 this issue is resolved.

Repository for reproducing the issue: https://github.com/woodylaurence/ScrutorAzureFunctionsExample.

### LINKED ISSUE #200: Duplicates when Scanning Assemblies with latest .NET 7 SDK versions
Hi,
We are using Scrutor as part of our .NET Core Web Api implementation.  The following code is used to scan dependencies of the api and register implementations of the handler interfaces:

`internal static IServiceCollection AutoRegisterHandlers(this IServiceCollection services)
{
    services.Scan(scan => scan.FromApplicationDependencies()
        .AddClasses(classes => classes.AssignableTo<IHandleGetRequests>()).As<IHandleGetRequests>()
        .AddClasses(classes => classes.AssignableTo<IHandleDeleteRequests>()).As<IHandleDeleteRequests>()
        .AddClasses(classes => classes.AssignableTo(typeof(IHandlePatchRequests<>))).AsImplementedInterfaces()
        .AddClasses(classes => classes.AssignableTo(typeof(IHandlePutRequests<>))).AsImplementedInterfaces()
        .AddClasses(classes => classes.AssignableTo(typeof(IHandlePostRequests<>))).AsImplementedInterfaces());
    return services;
}`

With version 7.0.100 of the .NET SDK the above code works as expected and the results of the call to services.Scan include one entry for each implementation of the interfaces.  However after upgrading the SDK to 7.0.2** (e.g. the latest version 7.0.203) then the assembly scanning produces duplicates for each interface, this in turn creates duplicates in our registration of the handlers and results in `Microsoft.AspNetCore.Routing.Matching.AmbiguousMatchException: The request matched multiple endpoints` errors at runtime.

We are aware of the RegistrationStrategy stuff and we are looking at amending our code to make use of this with either Skip or Replace but wanted to check if you are aware of the issue and whether there are any plans to release an update of Scrutor in response to this or not?

### LINKED ISSUE #208: Scrutor stopped working after upgrading to .NET 8 preview 7 from .NET 8 preview 6
The following code stopped working after upgrading from .NET 8 preview 6 to .NET 8 preview 7.

```csproj
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework> <!-- works well on net7.0 and net 8.0 preview 6, breaks on preview 7 -->
    <LangVersion>preview</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Scrutor" Version="4.2.2" />
  </ItemGroup>

</Project>
```

```csharp
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;

var builder = WebApplication.CreateBuilder(args);

builder.Services
    .AddTransient<IService, Service>()
    .AddTransient<IDecorated, Decorated1>()
    .Decorate<IDecorated, Decorated2>();

var app = builder.Build();

app.Map("/", (IService s) => s.ToString());

app.UseRouting();

app.Run();

interface IService;

class Service(IDecorated decorated) : IService;

interface IDecorated;

class Decorated1 : IDecorated;

class Decorated2(IDecorated decorated) : IDecorated;
```

```json
{
  "sdk": {
    "version": "8.0.100-preview.7.23376.3" // global.json
  }
}
```

With this exception

```
Unhandled exception. System.AggregateException: Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: IService Lifetime: Transient ImplementationType: Service': Unable to resolve service for type 'IDecorated' while attempting to activate 'Service'.)
 ---> System.InvalidOperationException: Error while validating the service descriptor 'ServiceType: IService Lifetime: Transient ImplementationType: Service': Unable to resolve service for type 'IDecorated' while attempting to activate 'Service'.
 ---> System.InvalidOperationException: Unable to resolve service for type 'IDecorated' while attempting to activate 'Service'.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateArgumentCallSites(ServiceIdentifier serviceIdentifier, Type implementationType, CallSiteChain callSiteChain, ParameterInfo[] parameters, Boolean throwIfCallSiteNotFound)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateConstructorCallSite(ResultCache lifetime, ServiceIdentifier serviceIdentifier, Type implementationType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateExact(ServiceDescriptor descriptor, ServiceIdentifier serviceIdentifier, CallSiteChain callSiteChain, Int32 slot)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.GetCallSite(ServiceDescriptor serviceDescriptor, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.ValidateService(ServiceDescriptor descriptor)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.ValidateService(ServiceDescriptor descriptor)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.Hosting.HostApplicationBuilder.Build()
   at Microsoft.AspNetCore.Builder.WebApplicationBuilder.Build()
   at Program.<Main>$(String[] args) in C:\Users\i\hate\devlooped\SponsorLink\source\repos\but\love\scrutor\Program.cs:line 11
```

### PR DESCRIPTION
**Pull request by khellang**
What do you think @DanHarltey?

Closes #171
Closes #208 
Closes #65
Closes #200

### CODE REVIEW

File: src/Scrutor/ServiceDescriptorExtensions.cs [hunk 1]
```diff
@@ -5,14 +5,46 @@ namespace Scrutor;
...
+        }
+
+        if (descriptor.ImplementationInstance is not null)
+        {
+            return new ServiceDescriptor(descriptor.ServiceType, serviceKey, descriptor.ImplementationInstance);
+        }
+
+        if (descriptor.ImplementationFactory is not null)
+        {
+            return new ServiceDescriptor(descriptor.ServiceType, serviceKey, (sp, key) => descriptor.ImplementationFactory(sp), descriptor.Lifetime);
```
Comment (DanHarltey) [inline src/Scrutor/ServiceDescriptorExtensions.cs:45]: Is it worth something like 
`            var implementationFactory = descriptor.ImplementationFactory;
            return new ServiceDescriptor(descriptor.ServiceType, serviceKey, (sp, key) => implementationFactory(sp), descriptor.Lifetime);`
So we don't catcher the descriptor in the closure?

**Review by DanHarltey — APPROVED** [review]
Nice. A lot nicer than the DecoratedType file, it is nasty.  Surprised by how little of the code actually changes. This should allow more compatibility.

**Comment by khellang** [comment]
Thanks for the review, @DanHarltey!  All good feedback 👍🏻

**Comment by dario-l** [comment]
Hi @khellang 
When this will be available at the nuget package? We really need this. 😃 

Please, please, please :)

**Comment by khellang** [comment]
I can probably target RC2 and publish a prerelease version if anyone's eager to test it out.

**Comment by dario-l** [comment]
> I can probably target RC2 and publish a prerelease version if anyone's eager to test it out.

Yes please 🙏🏻

**Comment by dario-l** [comment]
@khellang Is there any possibility to Scan and register types with provided key?

**Comment by dario-l** [comment]
Hi @khellang 😃 
I'm just cloned branch for net8 and adopted source code into my project. 😄 

Now I can wait for released version of net8 and newest scrutor 😃 
I hope that Scan will have appropriate extensions for keyed services also. 👍🏻 
Thanks for help. ❤️

**Comment by khellang** [comment]
> I hope that Scan will have appropriate extensions for keyed services also. 👍🏻

What does that mean? How do you envision using keyed services with Scrutor?

**Comment by dario-l** [comment]
> > I hope that Scan will have appropriate extensions for keyed services also. 👍🏻
> 
> What does that mean? How do you envision using keyed services with Scrutor?

I would love to register services from given assembly as keyed with specified key,

For example. I have modular monolith. Each module is a separated project/assembly. I execute `services.Scan(...)` to register all `ICommandHandler` from that assembly. I want to register them as keyed services. Thanks to that I can resolve only those services in particular module.

For now I am using this hacky solution 😄 
```
        var tempServices = new ServiceCollection();

        tempServices.Scan(  );
        
        foreach (var service in tempServices)
        {
                // register service as keyed
        }
```

**Comment by khellang** [comment]
If it's allowed to register multiple services against the same key, it could work. I'll have to verify it.

**Comment by dario-l** [comment]
Hi @khellang :)
How about this? Will be any chance for this? 😃

**Comment by dario-l** [comment]
Hi @khellang ?

How about now, after NET8 premiere? 😄

**Comment by khellang** [comment]
@rekosko What exactly are you waiting for? The latest version on NuGet works fine on .NET 8, right?

**Comment by rekosko** [comment]
@khellang yes I confirm it works on .NET 8 :) But decoration for keyed services is not yet implemented or I missed the purpose of this PR?

**Comment by khellang** [comment]
That's not the purpose of this PR, no. Although I want to add that as well before shipping a new version. This PR replaces an old hack to avoid StackOverflowExceptions when decorating services with keyed services under the covers. It also files a few other long-standing issues and pending breaking changes.

**Comment by khellang** [comment]
BTW, if anyone wants to contribute support for decorating keyed services to get a release out sooner, that would be cool.

**Comment by savornicesei** [comment]
@khellang Can this be merged into master or it's not complete or it needs more testing?

**Comment by khellang** [comment]
It's not totally complete and needs more tests. It's better to base work off this branch for now. Once this goes into master, I will release a new major version.

**Comment by abatishchev** [comment]
Gentle ping on this. I wonder what's the plan/fate for this PR?

**Comment by spko** [comment]
Howdy! Will Scrutor be ported to .NET 8? 
.NET 6 is getting very close to its End of Support deadline.

**Comment by khellang** [comment]
> Will Scrutor be ported to .NET 8? .NET 6 is getting very close to its End of Support deadline.

Just because the library targets .NET 6, doesn't mean it doesn't support .NET 8. Have you encountered any issues using the package on .NET 8?

**Comment by spko** [comment]
> Just because the library targets .NET 6, doesn't mean it doesn't support .NET 8. It's perfectly fine to use packages targeting older framework versions on newer frameworks.
> Have you encountered any issues using the package on .NET 8?

No, just curious, we are considering using Scrutor in one of our projects, but stricter corporate policies prevent us from using packages targetting .NET 6 🤷

**Comment by khellang** [comment]
> No, just curious, we are considering using Scrutor in one of our projects, but stricter corporate policies prevent us from using packages targetting .NET 6 🤷

It's obvious that someone in your company doesn't have a clue how .NET works - unfortunately I can't help with that 😅 

What version a package targets it 100% irrelevant to support policies - it's the version of your runtime that matters. If you're running .NET 8 using a package that targets anything newer than .NET Core 2, you're totally supported.

I'm curious; how do you enforce that? 🤔 

I can't really commit to anything, as I have a lot going on in my life right now and the package works fine on .NET 8 as-is. I'll do my best to get this PR in and ship .NET 8 support when I can find the time.

**Comment by 121GWJolt** [comment]
Understandable; life is a priority, as it ought to be.

I have not contributed to this project before, but would like to know if there is any way for someone to possibly assist so you don't have to do what life won't permit you to have time for. I don't know what I'll have time for either, but if there's a way to possibly help get this over the line, I'd like to help.

**Comment by khellang** [comment]
> would like to know if there is any way for someone to possibly assist so you don't have to do what life won't permit you to have time for

Honestly, I could probably just merge this PR and release v5.0 and then follow up with any bug fixes. What's lacking is tests for named services. If anyone wants to contribute some tests for named services, that'd be stellar. Otherwise I'll probably go ahead and just merge this and do a release and see how it goes 😅

**Comment by khellang** [comment]
Everyone, I've gone ahead and merged this and released [v5.0.0](https://www.nuget.org/packages/Scrutor/5.0.0) to NuGet. Please try it out and report back if you find any issues 🙏🏻

**Comment by ThumbGen** [comment]
Is there an example somewhere how to perform 'keyed registrations' using the latest Scrutor version? TIA

**Comment by khellang** [comment]
As mentioned in https://github.com/khellang/Scrutor/pull/209#issuecomment-1888728921, this isn't adding support for keyed registrations, it's merely (ab)using them internally to avoid StackOverflowExceptions when doing decoration :smile:

If someone is up to the task to implement support for keyed registrations, I'd love to see it ❤️

Do you have any thoughts on what it would look like, @ThumbGen?

**Comment by ThumbGen** [comment]
Thanks for the fast reply. Sadly I have no thoughts on it, I am just an avid consumer of this lovely library :-(

**Comment by julealgon** [comment]
@khellang I saw this PR mentioned in [the 5.0.1 release notes](https://github.com/khellang/Scrutor/releases/tag/v5.0.1) and just wanted to make sure I get this right: this change is just to rely on keyed service capabilities internally, but it results in **_the exact same behavior as before, including all the limitations that existed before_**. Is that a valid way of putting it?

If not, could you elaborate on what this could end up changing or what limitations it removes in terms of using decorators?

**Comment by khellang** [comment]
> Is that a valid way of putting it?

@julealgon That's correct, yes. Earlier, we relied internally on a huge hack to make decoration work, which ended up breaking a few scenarios, like Azure Functions. This PR replace that solution with a less hacky solution based on keyed services.

There is a potential benefit of this, which is that you could resolved a keyed service to get a hold of the "original", wrapped, service, but that hasn't been tested fully.

**Pull request merged**